### PR TITLE
chore(refactor): Refactor the /token endpoint to improve clarity.

### PR DESCRIPTION
**Pull request by rfk**
Continues https://github.com/mozilla/fxa-auth-server-private/pull/90 in the public repo.

Here's my attempt at a broader refactor of the /token route, as described in https://bugzilla.mozilla.org/show_bug.cgi?id=1531648. There are longer functions, but the control flow is much simpler IMHO and therefore easier to follow.

### CODE REVIEW

File: fxa-oauth-server/lib/db/memory.js [hunk 1]
```diff
@@ -155,8 +155,8 @@ MemoryStore.prototype = {
     client.canGrant = !! client.canGrant;
     client.trusted = !! client.trusted;
     this.clients[hex] = client;
-    client.hashedSecret = client.hashedSecret;
-    client.hashedSecretPrevious = client.hashedSecretPrevious || '';
+    client.hashedSecret = client.hashedSecret ? buf(client.hashedSecret) : null;
+    client.hashedSecretPrevious = client.hashedSecretPrevious ? buf(client.hashedSecretPrevious) : null;
```
Comment (rfk) [inline fxa-oauth-server/lib/db/memory.js]: The memory db was sometimes returning hex strings and sometimes buffers for these fields; this makes them consistently buffers, just like the mysql backend would produce.

File: fxa-oauth-server/lib/routes/token.js [hunk 2]
```diff
@@ -139,221 +146,199 @@ module.exports = {
...
-        throw AppError.expiredCode(code, expiresAt);
-      }
-    }
-    return db.removeCode(buf(code)).then(function() {
-      return codeObj;
+async function validateRefreshTokenGrant(client, params) {
+  // Does the refresh token actually exist?
+  const tokObj = await db.getRefreshToken(encrypt.hash(params.refresh_token));
+  if (! tokObj) {
+    logger.debug('refresh_token.notFound', params.refresh_token);
```
Comment (shane-tomlinson) [inline fxa-oauth-server/lib/routes/token.js]: For Stackdriver to injest as expected, does the 2nd parameter to all logger methods need to be an object? cc @jbuck

File: fxa-oauth-server/lib/routes/token.js [hunk 3]
```diff
@@ -139,221 +146,199 @@ module.exports = {
...
+  switch (params.grant_type) {
+  case GRANT_AUTHORIZATION_CODE:
+    requestedGrant = await validateAuthorizationCodeGrant(client, params);
+    break;
+  case GRANT_REFRESH_TOKEN:
+    requestedGrant = await validateRefreshTokenGrant(client, params);
+    break;
+  default:
+    // Joi validation means this should never happen.
+    logger.critical('joi.grant_type', params.grant_type);
```
Comment (shane-tomlinson) [inline fxa-oauth-server/lib/routes/token.js]: Same question here about Stackdriver logging and whether the 2nd param needs to be an object.

File: fxa-oauth-server/lib/routes/token.js [hunk 4]
```diff
@@ -375,43 +360,30 @@ function generateIdToken(options, access) {
...
+  };
+  result.expires_in = options.ttl;
+  if (options.authAt) {
+    result.auth_at = options.authAt;
+  }
+  if (options.keysJwe) {
+    result.keys_jwe = options.keysJwe;
+  }
+  // Maybe also generate a refreshToken?
+  if (options.offline) {
```
Comment (shane-tomlinson) [inline fxa-oauth-server/lib/routes/token.js]: If a refreshToken was used to generate this token, will a new refreshToken be generated? From the [db schema](https://github.com/mozilla/fxa-auth-server/blob/40b65eeb05a11437ae32cbf4b2ed06e30d71f261/fxa-oauth-server/lib/db/mysql/schema.sql#L74) it doesn't look like that's possible. Just want to make sure.

Comment (rfk) [reply fxa-oauth-server/lib/routes/token.js]: It shouldn't be possible, but we should probably make that more explicit.

File: fxa-oauth-server/lib/routes/token.js [hunk 5]
```diff
@@ -139,221 +146,199 @@ module.exports = {
...
+    db: stored,
+    dbPrevious: storedPrevious
+  });
+  throw AppError.incorrectSecret(client.id);
+}
 
-  return true;
+
+async function getClientById(clientId) {
+  const client = db.getClient(buf(clientId));
```
[the remark this answers is not in the document]
Comment (rfk) [reply fxa-oauth-server/lib/routes/token.js]: Nice catch! I'm surprised this didn't trigger a test failure, will see if I can add a failing test.

File: fxa-oauth-server/lib/routes/token.js (hunk 2 above)
Comment (rfk) [reply fxa-oauth-server/lib/routes/token.js]: I'll fix them for consistency, even if they might haved worked as-is.

**Comment by rfk** [comment]
Updated, with an additional test that failed with the missing `await` and is passing now that it's present.

File: fxa-oauth-server/lib/routes/token.js (hunk 2 above)
Comment (jbuck) [reply fxa-oauth-server/lib/routes/token.js]: How does mozlog transform it? Without knowing that, I'd assume it's safer to convert it into an object yourself

**Pull request merged**